src/mscorlib/src/System/Collections/Generic/EqualityComparer.cs

```diff
@@ -33,49 +33,73 @@ public abstract class EqualityComparer<T> : IEqualityComparer, IEqualityComparer
-            if (t.IsEnum) {
+            else if (t.IsEnum) {
                 TypeCode underlyingTypeCode = Type.GetTypeCode(Enum.GetUnderlyingType(t));
 
                 // Depending on the enum type, we need to special case the comparers so that we avoid boxing
                 // Note: We have different comparers for Short and SByte because for those types we need to make sure we call GetHashCode on the actual underlying type as the 
                 // implementation of GetHashCode is more complex than for the other types.
                 switch (underlyingTypeCode) {
                     case TypeCode.Int16: // short
-                        return (EqualityComparer<T>)RuntimeTypeHandle.CreateInstanceForAnotherGenericParameter((RuntimeType)typeof(ShortEnumEqualityComparer<short>), t);
+                        result = RuntimeTypeHandle.CreateInstanceForAnotherGenericParameter((RuntimeType)typeof(ShortEnumEqualityComparer<short>), t);
+                        break;
                     case TypeCode.SByte:
-                        return (EqualityComparer<T>)RuntimeTypeHandle.CreateInstanceForAnotherGenericParameter((RuntimeType)typeof(SByteEnumEqualityComparer<sbyte>), t);
+                        result = RuntimeTypeHandle.CreateInstanceForAnotherGenericParameter((RuntimeType)typeof(SByteEnumEqualityComparer<sbyte>), t);
+                        break;
                     case TypeCode.Int32:
                     case TypeCode.UInt32:
                     case TypeCode.Byte:
                     case TypeCode.UInt16: //ushort
-                        return (EqualityComparer<T>)RuntimeTypeHandle.CreateInstanceForAnotherGenericParameter((RuntimeType)typeof(EnumEqualityComparer<int>), t);
+                        result = RuntimeTypeHandle.CreateInstanceForAnotherGenericParameter((RuntimeType)typeof(EnumEqualityComparer<int>), t);
+                        break;
                     case TypeCode.Int64:
                     case TypeCode.UInt64:
-                        return (EqualityComparer<T>)RuntimeTypeHandle.CreateInstanceForAnotherGenericParameter((RuntimeType)typeof(LongEnumEqualityComparer<long>), t);
+                        result = RuntimeTypeHandle.CreateInstanceForAnotherGenericParameter((RuntimeType)typeof(LongEnumEqualityComparer<long>), t);
+                        break;
+                    default:
+                        Contract.Assert(false, "The enum's underlying TypeCode doesn't match any of the primitive types!");
+                        break;
                 }
             }
-            // Otherwise return an ObjectEqualityComparer<T>
+            
+            if (result != null)
+            {
+                Contract.Assert(result is EqualityComparer<T>,
+                    $"{nameof(result)} should always be a EqualityComparer if we reach here!");
+                return JitHelpers.UnsafeCast<EqualityComparer<T>>(result);
```

Comment:
The Unsafe cast should be only used in places where it has measurable benefits. I do not think it has any measurable benefit here.

This is a one-time initialization. The size of code matters more than a few extra cycles for one-time initialization. The regular cast should generate slightly less code overall because of it does not require method instantiation. Check the size of System.Private.CoreLib.ni.dll with regular cast here vs. unsafe cast here - the regular cast should result into smaller binary size.